File: go/stduritemplate.go
package stduritemplate

import (
	"fmt"
	"net/url"
	"sort"
	"strconv"
	"strings"
	"time"
)

type Substitutions map[string]any

type Modifier int

const (
	ModUndefined Modifier = iota
	ModNone
	ModPlus
	ModDash
	ModDot
	ModSlash
	ModSemicolon
	ModQuestionMark
	ModAt
)

const (
	SubstitutionTypeString = "STRING"
	SubstitutionTypeList   = "LIST"
	SubstitutionTypeMap    = "MAP"
)

func validateLiteral(c rune, col int) error {
	switch c {
	case '+', '#', '/', ';', '?', '&', ' ', '!', '=', '$', '|', '*', ':', '~', '-':
		return fmt.Errorf("illegal character identified in the token at col: %d", col)
	default:
		return nil
	}
}

func getMaxChar(buffer *strings.Builder, col int) (int, error) {
	if buffer == nil {
		return -1, nil
	}
	value := buffer.String()

	if value == "" {
		return -1, nil
	}

	maxChar, err := strconv.Atoi(value)
	if err != nil {
		return 0, fmt.Errorf("cannot parse max chars at col: %d", col)
	}
	return maxChar, nil
}

func getModifier(c rune, token *strings.Builder, col int) (Modifier, error) {
	switch c {
	case '+':
		return ModPlus, nil
	case '#':
		return ModDash, nil
	case '.':
		return ModDot, nil
	case '/':
		return ModSlash, nil
	case ';':
		return ModSemicolon, nil
	case '?':
		return ModQuestionMark, nil
	case '&':
		return ModAt, nil
	default:
		err := validateLiteral(c, col)
		if err != nil {
			return ModUndefined, err
		}
		token.WriteRune(c)
		return ModNone, nil
	}
}

func expandImpl(str string, substitutions Substitutions) (string, error) {
	var result strings.Builder

	var token *strings.Builder
	var modifier = ModUndefined
	var composite bool
	var maxCharBuffer *strings.Builder
	var firstToken = true

	for i, character := range str {
		switch character {
		case '{':
			token = &strings.Builder{}
			firstToken = true
		case '}':
			if token != nil {
				maxChar, err := getMaxChar(maxCharBuffer, i)
				if err != nil {
					return "", err
				}
				expanded, err := expandToken(modifier, token.String(), composite, maxChar, firstToken, substitutions, &result, i)
				if err != nil {
					return "", err
				}
				if expanded && firstToken {
					firstToken = false
				}
				token = nil
				modifier = ModUndefined
				composite = false
				maxCharBuffer = nil
			} else {
				return "", fmt.Errorf("failed to expand token, invalid at col: %d", i)
			}
		case ',':
			if token != nil {
				maxChar, err := getMaxChar(maxCharBuffer, i)
				if err != nil {
					return "", err
				}
				expanded, err := expandToken(modifier, token.String(), composite, maxChar, firstToken, substitutions, &result, i)
				if err != nil {
					return "", err
				}
				if expanded && firstToken {
					firstToken = false
				}
				token = &strings.Builder{}
				composite = false
				maxCharBuffer = nil
				break
			}
			// Intentional fall-through for commas outside the {}
			fallthrough
		default:
			if token != nil {
				switch {
				case modifier == ModUndefined:
					var err error
					modifier, err = getModifier(character, token, i)
					if err != nil {
						return "", err
					}
				case maxCharBuffer != nil:
					if _, err := strconv.Atoi(string(character)); err == nil {
						maxCharBuffer.WriteRune(character)
					} else {
						return "", fmt.Errorf("illegal character identified in the token at col: %d", i)
					}
				default:
					switch character {
					case ':':
						maxCharBuffer = &strings.Builder{}
					case '*':
						composite = true
					default:
						if err := validateLiteral(character, i); err != nil {
							return "", err
						}
						token.WriteRune(character)
					}
				}
			} else {
				result.WriteRune(character)
			}
		}
	}

	if token == nil {
		return result.String(), nil
	}

	return "", fmt.Errorf("unterminated token")
}

func addPrefix(mod Modifier, result *strings.Builder) {
	switch mod {
	case ModDash:
		result.WriteByte('#')
	case ModDot:
		result.WriteByte('.')
	case ModSlash:
		result.WriteByte('/')
	case ModSemicolon:
		result.WriteByte(';')
	case ModQuestionMark:
		result.WriteByte('?')
	case ModAt:
		result.WriteByte('&')
	default:
		return
	}
}

func addSeparator(mod Modifier, result *strings.Builder) {
	switch mod {
	case ModDot:
		result.WriteByte('.')
	case ModSlash:
		result.WriteByte('/')
	case ModSemicolon:
		result.WriteByte(';')
	case ModQuestionMark, ModAt:
		result.WriteByte('&')
	default:
		result.WriteByte(',')
		return
	}
}

func addValue(mod Modifier, token, value string, result *strings.Builder, maxChar int) {
	switch mod {
	case ModPlus, ModDash:
		addExpandedValue(value, result, maxChar, false)
	case ModQuestionMark, ModAt:
		result.WriteString(token + "=")
		addExpandedValue(value, result, maxChar, true)
	case ModSemicolon:
		result.WriteString(token)
		if value != "" {
			result.WriteByte('=')
		}
		addExpandedValue(value, result, maxChar, true)
	case ModDot, ModSlash, ModNone:
		addExpandedValue(value, result, maxChar, true)
	}
}

func addValueElement(mod Modifier, _, value string, result *strings.Builder, maxChar int) {
	switch mod {
	case ModPlus, ModDash:
		addExpandedValue(value, result, maxChar, false)
	case ModQuestionMark, ModAt, ModSemicolon, ModDot, ModSlash, ModNone:
		addExpandedValue(value, result, maxChar, true)
	}
}

func addExpandedValue(value string, result *strings.Builder, maxChar int, replaceReserved bool) {
	max := maxChar
	if maxChar == -1 || maxChar > len(value) {
		max = len(value)
	}
	reservedBuffer := &strings.Builder{}
	fillReserved := false

	for i, character := range value {
		if i >= max {
			break
		}

		if character == '%' && !replaceReserved {
			reservedBuffer.Reset()
			fillReserved = true
		}

		if fillReserved {
			reservedBuffer.WriteRune(character)

			if reservedBuffer.Len() == 3 {
				encoded := true
				reserved := reservedBuffer.String()
				unescaped, err := url.QueryUnescape(reserved)
				if err != nil {
					encoded = (reserved == unescaped)
				}

				if encoded {
					result.WriteString(reserved)
				} else {
					result.WriteString("%25")
					// only if !replaceReserved
					result.WriteString(reservedBuffer.String()[1:])
				}
				reservedBuffer.Reset()
				fillReserved = false
			}
		} else {
			switch character {
			case ' ':
				result.WriteString("%20")
			case '%':
				result.WriteString("%25")
			default:
				if replaceReserved {
					result.WriteString(url.QueryEscape(string(character)))
				} else {
					result.WriteRune(character)
				}
			}
		}
	}

	if fillReserved {
		result.WriteString("%25")
		if replaceReserved {
			result.WriteString(url.QueryEscape(reservedBuffer.String()[1:]))
		} else {
			result.WriteString(reservedBuffer.String()[1:])
		}
	}
}

func getSubstitutionType(value any, col int) string {
	switch value.(type) {
	case string, nil:
		return SubstitutionTypeString
	case []any:
		return SubstitutionTypeList
	case map[string]any:
		return SubstitutionTypeMap
	default:
		return fmt.Sprintf("illegal class passed as substitution, found %T at col: %d", value, col)
	}
}

func isEmpty(substType string, value any) bool {
	switch substType {
	case SubstitutionTypeString:
		return value == nil
	case SubstitutionTypeList:
		return len(value.([]any)) == 0
	case SubstitutionTypeMap:
		return len(value.(map[string]any)) == 0
	default:
		return true
	}
}

func expandToken(
	modifier Modifier,
	token string,
	composite bool,
	maxChar int,
	firstToken bool,
	substitutions Substitutions,
	result *strings.Builder,
	col int,
) (bool, error) {
	if len(token) == 0 {
		return false, fmt.Errorf("found an empty token at col: %d", col)
	}

	value, ok := substitutions[token]
	if !ok {
		return false, nil
	}

	switch value.(type) {
	case bool, int, int64, float32, float64:
		value = fmt.Sprintf("%v", value)
	case time.Time:
		value = value.(time.Time).Format(time.RFC3339)
	}

	substType := getSubstitutionType(value, col)
	if isEmpty(substType, value) {
		return false, nil
	}

	if firstToken {
		addPrefix(modifier, result)
	} else {
		addSeparator(modifier, result)
	}

	switch substType {
	case SubstitutionTypeString:
		addStringValue(modifier, token, value.(string), result, maxChar)
	case SubstitutionTypeList:
		addListValue(modifier, token, value.([]any), result, maxChar, composite)
	case SubstitutionTypeMap:
		err := addMapValue(modifier, token, value.(map[string]any), result, maxChar, composite)
		if err != nil {
			return false, err
		}
	}

	return true, nil
}

func addStringValue(modifier Modifier, token string, value string, result *strings.Builder, maxChar int) {
	addValue(modifier, token, value, result, maxChar)

}

func addListValue(modifier Modifier, token string, value []any, result *strings.Builder, maxChar int, composite bool) {
	first := true
	for _, v := range value {
		if first {
			addValue(modifier, token, v.(string), result, maxChar)
			first = false
		} else {
			if composite {
				addSeparator(modifier, result)
				addValue(modifier, token, v.(string), result, maxChar)
			} else {
				result.WriteString(",")
				addValueElement(modifier, token, v.(string), result, maxChar)
			}
		}
	}
}

func addMapValue(modifier Modifier, token string, value map[string]any, result *strings.Builder, maxChar int, composite bool) error {
	first := true
	if maxChar != -1 {
		return fmt.Errorf("value trimming is not allowed on Maps")
	}

	// workaround to make Map ordering not random
	// https://github.com/uri-templates/uritemplate-test/pull/58#issuecomment-1640029982
	keys := make([]string, 0, len(value))
	for k := range value {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for key := range keys {
		k := keys[key]
		v := value[k]

		if composite {
			if !first {
				addSeparator(modifier, result)
			}
			addValueElement(modifier, token, k, result, maxChar)
			result.WriteString("=")
		} else {
			if first {
				addValue(modifier, token, k, result, maxChar)
			} else {
				result.WriteString(",")
				addValueElement(modifier, token, k, result, maxChar)
			}
			result.WriteString(",")
		}
		addValueElement(modifier, token, v.(string), result, maxChar)
		first = false
	}
	return nil
}

func Expand(template string, substitutions Substitutions) (string, error) {
	return expandImpl(template, substitutions)
}

File: java/src/main/java/io/github/stduritemplate/StdUriTemplate.java
package io.github.stduritemplate;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class StdUriTemplate {

    // Public API
    public static String expand(final String template, final Map<String, Object> substitutions) {
        return expandImpl(template, substitutions);
    }

    // Private implementation
    private enum Operator {
        NO_OP,
        PLUS,
        HASH,
        DOT,
        SLASH,
        SEMICOLON,
        QUESTION_MARK,
        AMP;
    }

    private static void checkVarname(String token, int col) {
        if (token.endsWith(".")) {
            throw new IllegalArgumentException("Variable name cannot end with a dot at col:" + col);
        }
        if (token.contains("..")) {
            throw new IllegalArgumentException("Variable name cannot contain consecutive dots at col:" + col);
        }
        for (int i = 0; i < token.length(); i++) {
            if (token.charAt(i) == '%') {
                if (i + 2 >= token.length()
                        || !isHexDigit(token.charAt(i + 1))
                        || !isHexDigit(token.charAt(i + 2))) {
                    throw new IllegalArgumentException("Invalid percent-encoding in variable name at col:" + col);
                }
            }
        }
    }

    private static boolean isHexDigit(char c) {
        return (c >= '0' && c <= '9') || (c >= 'A' && c <= 'F') || (c >= 'a' && c <= 'f');
    }

    private static void validateLiteral(Character c, int col) {
        switch (c) {
            case '+':
            case '#':
            case '/':
            case ';':
            case '?':
            case '&':
            case ' ':
            case '!':
            case '=':
            case '$':
            case '|':
            case '*':
            case ':':
            case '~':
            case '-':
                throw new IllegalArgumentException("Illegal character identified in the token at col:" + col);
            default:
                break;
        }
    }

    private static int getMaxChar(StringBuilder buffer, int col) {
        if (buffer == null || buffer.length() == 0) {
            return -1;
        } else {
            String value = buffer.toString();

            if (value.isEmpty()) {
                return -1;
            } else {
                try {
                    if (value.charAt(0) == '0') {
                        throw new IllegalArgumentException("Leading zero not allowed in max chars at col:" + col);
                    }
                    int result = Integer.parseInt(value);
                    if (result < 1 || result > 9999) {
                        throw new IllegalArgumentException("Max chars out of range (1-9999) at col:" + col);
                    }
                    return result;
                } catch (NumberFormatException e) {
                    throw new IllegalArgumentException("Cannot parse max chars at col:" + col);
                }
            }
        }
    }

    private static Operator getOperator(Character c, StringBuilder token, int col) {
        switch (c) {
            case '+': return Operator.PLUS;
            case '#': return Operator.HASH;
            case '.': return Operator.DOT;
            case '/': return Operator.SLASH;
            case ';': return Operator.SEMICOLON;
            case '?': return Operator.QUESTION_MARK;
            case '&': return Operator.AMP;
            default:
                validateLiteral(c, col);
                token.append(c);
                return Operator.NO_OP;
        }
    }

    private static String expandImpl(String str, Map<String, Object> substitutions) {
        final StringBuilder result = new StringBuilder(str.length() * 2);

        boolean toToken = false;
        final StringBuilder token = new StringBuilder();

        Operator operator = null;
        boolean composite = false;
        boolean toMaxCharBuffer = false;
        final StringBuilder maxCharBuffer = new StringBuilder(3);
        boolean firstToken = true;

        for (int i = 0; i < str.length(); i++) {
            char character = str.charAt(i);
            switch (character) {
                case '{':
                    toToken = true;
                    token.setLength(0);
                    firstToken = true;
                    break;
                case '}':
                    if (toToken) {
                        if (toMaxCharBuffer && maxCharBuffer.length() == 0) {
                            throw new IllegalArgumentException("Empty prefix modifier at col:" + i);
                        }
                        boolean expanded = expandToken(operator, token.toString(), composite, getMaxChar(maxCharBuffer, i), firstToken, substitutions, result, i);
                        if (expanded && firstToken) {
                            firstToken = false;
                        }
                        toToken = false;
                        token.setLength(0);
                        operator = null;
                        composite = false;
                        toMaxCharBuffer = false;
                        maxCharBuffer.setLength(0);
                    } else {
                        throw new IllegalArgumentException("Failed to expand token, invalid at col:" + i);
                    }
                    break;
                case ',':
                    if (toToken) {
                        if (toMaxCharBuffer && maxCharBuffer.length() == 0) {
                            throw new IllegalArgumentException("Empty prefix modifier at col:" + i);
                        }
                        boolean expanded = expandToken(operator, token.toString(), composite, getMaxChar(maxCharBuffer, i), firstToken, substitutions, result, i);
                        if (expanded && firstToken) {
                            firstToken = false;
                        }
                        token.setLength(0);
                        composite = false;
                        toMaxCharBuffer = false;
                        maxCharBuffer.setLength(0);
                        break;
                    }
                    // Intentional fall-through for commas outside the {}
                default:
                    if (toToken) {
                        if (operator == null) {
                            operator = getOperator(character, token, i);
                        } else if (toMaxCharBuffer) {
                            if (Character.isDigit(character)) {
                                maxCharBuffer.append(character);
                            } else {
                                throw new IllegalArgumentException("Illegal character identified in the token at col:" + i);
                            }
                        } else {
                            if (character == ':') {
                                toMaxCharBuffer = true;
                                maxCharBuffer.setLength(0);
                            } else if (character == '*') {
                                composite = true;
                            } else {
                                validateLiteral(character, i);
                                token.append(character);
                            }
                        }
                    } else {
                        if (character > 0x7F || isSurrogate(character)) {
                            try {
                                String toEncode;
                                if (isSurrogate(character) && i + 1 < str.length()) {
                                    toEncode = new String(new char[]{character, str.charAt(++i)});
                                } else {
                                    toEncode = Character.toString(character);
                                }
                                result.append(URLEncoder.encode(toEncode, StandardCharsets.UTF_8.name()));
                            } catch (UnsupportedEncodingException e) {
                                throw new RuntimeException(e);
                            }
                        } else {
                            result.append(character);
                        }
                    }
                    break;
            }
        }

        if (!toToken) {
            return result.toString();
        } else {
            throw new IllegalArgumentException("Unterminated token");
        }
    }

    private static void addPrefix(Operator op, StringBuilder result) {
        switch (op) {
            case HASH:
                result.append('#');
                break;
            case DOT:
                result.append('.');
                break;
            case SLASH:
                result.append('/');
                break;
            case SEMICOLON:
                result.append(';');
                break;
            case QUESTION_MARK:
                result.append('?');
                break;
            case AMP:
                result.append('&');
                break;
            default:
                return;
        }
    }

    private static void addSeparator(Operator op, StringBuilder result) {
        switch (op) {
            case DOT:
                result.append('.');
                break;
            case SLASH:
                result.append('/');
                break;
            case SEMICOLON:
                result.append(';');
                break;
            case QUESTION_MARK:
            case AMP:
                result.append('&');
                break;
            default:
                result.append(',');
                return;
        }
    }

    private static void addValue(Operator op, String token, Object value, StringBuilder result, int maxChar) {
        switch (op) {
            case PLUS:
            case HASH:
                addExpandedValue(null, value, result, maxChar, false);
                break;
            case QUESTION_MARK:
            case AMP:
                result.append(token + '=');
                addExpandedValue(null, value, result, maxChar, true);
                break;
            case SEMICOLON:
                result.append(token);
                addExpandedValue("=", value, result, maxChar, true);
                break;
            case DOT:
            case SLASH:
            case NO_OP:
                addExpandedValue(null, value, result, maxChar, true);
        }
    }

    private static void addValueElement(Operator op, String token, Object value, StringBuilder result, int maxChar) {
        switch (op) {
            case PLUS:
            case HASH:
                addExpandedValue(null, value, result, maxChar, false);
                break;
            case QUESTION_MARK:
            case AMP:
            case SEMICOLON:
            case DOT:
            case SLASH:
            case NO_OP:
                addExpandedValue(null, value, result, maxChar, true);
        }
    }

    private static boolean isSurrogate(char cp) {
        return (cp >= 0xD800 && cp <= 0xDFFF);
    }

    private static boolean isIprivate(char cp) {
        return (0xE000 <= cp && cp <= 0xF8FF);
    }

    private static boolean isUcschar(char cp) {
        return (0xA0 <= cp && cp <= 0xD7FF)
                || (0xF900 <= cp && cp <= 0xFDCF)
                || (0xFDF0 <= cp && cp <= 0xFFEF);
    }

    private static void addExpandedValue(String prefix, Object value, StringBuilder result, int maxChar, boolean replaceReserved) {
        String stringValue = convertNativeTypes(value);
        int cpCount = stringValue.codePointCount(0, stringValue.length());
        int max = (maxChar != -1) ? Math.min(maxChar, cpCount) : cpCount;
        result.ensureCapacity(max * 2); // hint to SB
        boolean toReserved = false;
        final StringBuilder reservedBuffer = new StringBuilder(3);

        if (max > 0 && prefix != null) {
            result.append(prefix);
        }

        int charCount = 0;
        for (int i = 0; i < stringValue.length() && charCount < max; i++) {
            char character = stringValue.charAt(i);
            charCount++;

            if (character == '%' && !replaceReserved) {
                toReserved = true;
                reservedBuffer.setLength(0);
            }

            String toAppend = Character.toString(character);
            try {
                if (isSurrogate(character)) {
                    StringBuilder sb = new StringBuilder();
                    sb.append(Character.toChars(stringValue.codePointAt(i++)));
                    toAppend = URLEncoder.encode(sb.toString(), StandardCharsets.UTF_8.name());
                } else if (replaceReserved || isUcschar(character) || isIprivate(character)) {
                    toAppend = URLEncoder.encode(toAppend, StandardCharsets.UTF_8.name());
                }
            } catch (UnsupportedEncodingException e) {
                throw new RuntimeException(e);
            }

            if (toReserved) {
                reservedBuffer.append(toAppend);

                if (reservedBuffer.length() == 3) {
                    boolean isEncoded = false;
                    try {
                        URLDecoder.decode(reservedBuffer.toString(), StandardCharsets.UTF_8.name());
                        isEncoded = true;
                    } catch (Exception e) {
                        // ignore
                    }

                    if (isEncoded) {
                        result.append(reservedBuffer);
                    } else {
                        result.append("%25");
                        // only if !replaceReserved
                        result.append(reservedBuffer.substring(1));
                    }
                    toReserved = false;
                    reservedBuffer.setLength(0);
                }
            } else {
                if (character == ' ') {
                    result.append("%20");
                } else if (character == '%') {
                    result.append("%25");
                } else {
                    result.append(toAppend);
                }
            }
        }

        if (toReserved) {
            result.append("%25");
            result.append(reservedBuffer.substring(1));
        }
    }

    private static boolean isList(Object value) {
        return value instanceof ArrayList || // checking concrete instances first as it's faster
                value instanceof List;
    }

    private static boolean isMap(Object value) {
        return value instanceof HashMap || // checking concrete instances first as it's faster
                value instanceof Map;
    }

    enum SubstitutionType {
        EMPTY,
        STRING,
        LIST,
        MAP;
    }

    private static SubstitutionType getSubstitutionType(Object value, int col) {
        if (value == null) {
            return SubstitutionType.EMPTY;
        } else if (isNativeType(value)) {
            return SubstitutionType.STRING;
        } else if (isList(value)) {
            return SubstitutionType.LIST;
        } else if (isMap(value)) {
            return SubstitutionType.MAP;
        } else {
            throw new IllegalArgumentException("Illegal class passed as substitution, found " + value.getClass() + " at col:" + col);
        }
    }

    private static boolean isEmpty(SubstitutionType substType, Object value) {
        if (value == null) {
            return true;
        } else {
            switch (substType) {
                case STRING: return false;
                case LIST: return ((List)value).isEmpty();
                case MAP: return ((Map)value).isEmpty();
                default: return true;
            }
        }
    }

    private static boolean isNativeType(Object value) {
        if (value instanceof String ||
            value instanceof Boolean ||
            value instanceof Integer ||
            value instanceof Long ||
            value instanceof Float ||
            value instanceof Double) {
            return true;
        }
        return false;
    }

    private static String convertNativeTypes(Object value) {
        if (value instanceof String) {
            return (String)value;
        } else if (value instanceof Boolean ||
            value instanceof Integer ||
            value instanceof Long ||
            value instanceof Float ||
            value instanceof Double) {
            return value.toString();
        }
        throw new IllegalArgumentException("Illegal class passed as substitution, found " + value.getClass());
    }

    // returns true if expansion happened
    private static boolean expandToken(
            Operator operator,
            String token,
            boolean composite,
            int maxChar,
            boolean firstToken,
            Map<String, Object> substitutions,
            StringBuilder result,
            int col) {
        if (token.isEmpty()) {
            throw new IllegalArgumentException("Found an empty token at col:" + col);
        }

        checkVarname(token, col);

        Object value = substitutions.get(token);
        SubstitutionType substType = getSubstitutionType(value, col);
        if (substType == SubstitutionType.EMPTY || isEmpty(substType, value)) {
            return false;
        }

        if (firstToken) {
            addPrefix(operator, result);
        } else {
            addSeparator(operator, result);
        }

        switch (substType) {
            case STRING:
                addStringValue(operator, token, (Object)value, result, maxChar);
                break;
            case LIST:
                addListValue(operator, token, (List<Object>)value, result, maxChar, composite);
                break;
            case MAP:
                addMapValue(operator, token, (Map<String, Object>)value, result, maxChar, composite);
                break;
        }

        return true;
    }

    private static boolean addStringValue(Operator operator, String token, Object value, StringBuilder result, int maxChar) {
        addValue(operator, token, value, result, maxChar);
        return true;
    }

    private static boolean addListValue(Operator operator, String token, List<Object> value, StringBuilder result, int maxChar, boolean composite) {
        boolean first = true;
        for (Object v: value) {
            if (first) {
                addValue(operator, token, v, result, maxChar);
                first = false;
            } else {
                if (composite) {
                    addSeparator(operator, result);
                    addValue(operator, token, v, result, maxChar);
                } else {
                    result.append(',');
                    addValueElement(operator, token, v, result, maxChar);
                }
            }
        }
        return !first;
    }

    private static boolean addMapValue(Operator operator, String token, Map<String, Object> value, StringBuilder result, int maxChar, boolean composite) {
        boolean first = true;
        if (maxChar != -1) {
            throw new IllegalArgumentException("Value trimming is not allowed on Maps");
        }
        for (Map.Entry<String, Object> v : value.entrySet()) {
            if (composite) {
                if (!first) {
                    addSeparator(operator, result);
                }
                addValueElement(operator, token, v.getKey(), result, maxChar);
                result.append('=');
            } else {
                if (first) {
                    addValue(operator, token, v.getKey(), result, maxChar);
                } else {
                    result.append(',');
                    addValueElement(operator, token, v.getKey(), result, maxChar);
                }
                result.append(',');
            }
            addValueElement(operator, token, v.getValue(), result, maxChar);
            first = false;
        }
        return !first;
    }

}
